compiler/src/iree/compiler/Dialect/HAL/IR/HALDialect.cpp

```diff
@@ -142,22 +142,38 @@ class HALAffinityAnalysisDialectInterface
       }
 
       MLIRContext *ctx = getContext();
+      std::optional<IREE::Encoding::UnsupportedEncodingAttr> defaultAttr;
+      auto getDefaultAttr = [&]() {
+        if (defaultAttr) {
+          return defaultAttr.value();
+        }
+        defaultAttr = IREE::Encoding::UnsupportedEncodingAttr::get(ctx);
+        return defaultAttr.value();
+      };
       for (IREE::Stream::AffinityAndOpPair key : batchQueries) {
         auto [affinityAttr, op] = key;
         SetVector<IREE::HAL::ExecutableTargetAttr> resultSet;
         deviceAnalysis.gatherRequiredExecutableTargets(affinityAttr, op,
                                                        resultSet);
         for (auto targetAttr : resultSet) {
-          Attribute result = IREE::Encoding::UnsupportedEncodingAttr::get(ctx);
-          if (auto attr = targetAttr.getConfiguration().getNamed("encoding")) {
-            if (auto encodingLayoutAttr =
-                    dyn_cast<IREE::Encoding::EncodingLayoutAttrInterface>(
-                        attr->getValue())) {
-              result = encodingLayoutAttr.cloneWithSimplifiedConfig(
-                  targetAttr.getConfiguration());
-            }
+          if (!targetAttr.hasConfigurationAttr("encoding")) {
```

Comment:
Nit: we should really be using constant variables instead of these floating strings imo.

Reply:
Yes, you're right. I've been thinking where should we put the constant variables because of dependencies. Now I think we can put it to `EncodingTypes.h` because they can all depend on the IREE::Encoding dialect. I'll do it in a follow-up.